### PR TITLE
Adding Helm-To-Operator-Codegen-Sdk Documentation

### PR DESCRIPTION
**Pull request by jain-ashish-sam**
The current pr contains the user-guide as well as developer-guide for helm-to-operator-codegen-sdk which is a part of nephio sdk (https://github.com/nephio-project/nephio-sdk)

### CODE REVIEW

File: content/en/docs/guides/user-guides/helm/_index.md [hunk 1]
```diff
@@ -16,6 +16,7 @@ for deploying on Kubernetes, legacy helm support will be required for a number o
 
 ### Deployment options
 
+* [Using Helm To Operator Codegen Sdk](/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md)
```
Comment (electrocucaracha) [inline content/en/docs/guides/user-guides/helm/_index.md:19]: I think that relative paths can work here
```suggestion
* [Using Helm To Operator Codegen Sdk](helm-to-operator-codegen-sdk-user-guide.md)
```

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
+---
+title: Helm to Operator Codegen Sdk Developer Guide
+weight: 1
+---
+
+The [Helm to Operator Codegen SDK](https://github.com/nephio-project/nephio-sdk/tree/main/helm-to-operator-codegen-sdk) offers a streamlined solution for translating existing Helm charts into Kubernetes operators with minimal effort and cost.
+
+## The Flow Diagram
+In a nutshell, Firstly, the Helm-Charts are converted to Yamls using the values provided in "values.yaml". Then, each Kubernetes Resource Manifest (KRM) in the YAML is translated into Go code, employing one of two methods.
```
Comment (electrocucaracha) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:9]: https://cloud.google.com/blog/topics/developers-practitioners/build-platform-krm-part-2-how-kubernetes-resource-model-works
```suggestion
In a nutshell, the Helm-Charts are converted to YAML files using the values provided in "values.yaml". Then, each Kubernetes Resource Model (KRM) in the YAML is translated into Go code, employing one of two methods.
```

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 3]
```diff
@@ -0,0 +1,333 @@
...
+1) If the resource is Runtime-Supported, it undergoes a conversion process where the KRM resource is first transformed into a Runtime Object, then into JSON, and finally into Go code.
+2) Otherwise, if the resource is not Runtime-Supported, it is converted into an Unstructured Object and then into Go code.
+
+After the conversion process, all the generated Go code is gathered and compiled into a single Go file. This resulting file contains functions that can be readily utilized by Kubernetes operators.
```
Comment (electrocucaracha) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:13]:
```suggestion
After the conversion process, all the generated Go code is gathered and compiled into a single Go file. This process results in a file that contains functions which can be readily utilized by Kubernetes operators.
```

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 4]
```diff
@@ -0,0 +1,333 @@
...
+
+![alt Flow Diagram](/static/images/user-guides/helm-to-operator-codegen-sdk-flow-diagram.jpg)
+
+-----
+### Flow-1: Helm to Yaml
+Helm to Yaml conversion is achieved by running the command
```
Comment (electrocucaracha) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:19]:
```suggestion
### Flow-1: Helm to YAML
Helm to YAML conversion is achieved by running the following command
```

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 5]
```diff
@@ -0,0 +1,333 @@
...
+`helm template <chart>  --namespace <namespace>  --output-dir “temp/templated/”` Internally. As of now, It retrieves the values from default "values.yaml"
```
Comment (electrocucaracha) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:20]:
```suggestion
`helm template <chart>  --namespace <namespace>  --output-dir “temp/templated/”` Internally. As of now, it retrieves the values from default "values.yaml"
```

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+
+### Flow-2: Yaml Split
```
Comment (electrocucaracha) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:22]:
```suggestion
### Flow-2: YAML Split
```

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 7]
```diff
@@ -0,0 +1,333 @@
...
+The SDK iterates over each YAML file in the "converted-yamls" directory. If a YAML file contains multiple Kubernetes Resource Manifests (KRM), separated by "---", the SDK splits the YAML file accordingly to isolate each individual KRM resource. This ensures that each KRM resource is processed independently.
```
Comment (electrocucaracha) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:23]:
```suggestion
The SDK iterates over each YAML file in the "converted-yamls" directory. If a YAML file contains multiple Kubernetes Resource Models (KRMs), separated by "---", the SDK splits the YAML file accordingly to isolate each individual KRM resource. This ensures that each KRM resource is processed independently.
```

**Review by electrocucaracha — COMMENTED** [review]
Initial review

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+By utilizing the Helm to Operator Codegen SDK, users can efficiently convert existing Helm charts into Kubernetes operators. This transition enables users to leverage the advanced capabilities provided by Kubernetes operators, such as enhanced automation, lifecycle management, and custom logic handling. Overall, the SDK facilitates a smooth migration process, empowering users to embrace the operator model for managing their Kubernetes resources effectively.
+
+## Excercise: Deploying Free5gc using operator
+In the following exercise, the objective is to convert the Free5gc Helm chart to Go code suitable for a Kubernetes operator using the sdk. Once the conversion is complete, the generated Go code will be used to deploy all the resources defined in the Free5gc Helm chart using a Kubernetes operator.
+
+### Step 0: Prerequisite
+1. GoLang Version: 1.21
+2. Helm : v3.9.3
+3. Kubebuilder
+4. A Kubernetes Cluster with Calico CNI and multus-cni plugin (Can Refer [here](https://medium.com/rahasak/deploying-5g-core-network-with-free5gc-kubernets-and-helm-charts-29741cea3922), Before "Deploy Helm-Chart Part" )
```
Comment (lapentad) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md:19]: Hi when using free5gc I always found it easier to work with [Flannel for this reason ](https://github.com/Orange-OpenSource/towards5gs-helm/blob/main/docs/demo/Setup-free5gc-and-test-with-UERANSIM.md#tun-interface-correctly-created-on-the-ue-but-internet)

> ipv4.ip_forward We remind you that some CNI plugins (e.g. [Flannel](https://github.com/flannel-io/flannel)) allow this functionality by default, while others (.e.g. [Calico](https://github.com/projectcalico/cni-plugin)) require a [special configuration](https://docs.projectcalico.org/reference/host-endpoints/forwarded).

**Comment by lapentad** [comment]
I think the Check links job is failing because the user guide is not in the _index.md. 

> Cannot find: helm-to-operator-codegen-sdk-user-guide.md in content/en/docs/guides/user-guides/helm/_index.md

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 9]
```diff
@@ -0,0 +1,333 @@
...
+```
+</details>
+
+### Flow-3.1: KRM to Runtime-Object
+The conversion process relies on the "k8s.io/apimachinery/pkg/runtime" package. Currently, only the API version "v1" is supported. The supported kinds for the Runtime Object method include:
+`Deployment, Service, Secret, Role, RoleBinding, ClusterRoleBinding, PersistentVolumeClaim, StatefulSet, ServiceAccount, ClusterRole, PriorityClass, ConfigMap`
+
+### Flow-3.2: Runtime-Object to JSON
+Firstly, the SDK performs a typecast of the runtime object to its actual data type. For instance, if the Kubernetes Kind is "Service," the SDK typecasts the runtime object to the specific data type corev1.Service. Then, it conducts a Depth-First Search (DFS) traversal over the corev1.Service object using reflection. During this traversal, the SDK generates a JSON structure that encapsulates information about the struct hierarchy, including corresponding data types and values. This transformation results in a JSON representation of the corev1.Service object's structure and content.
+<details>
```
Comment (CsatariGergely) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:79]: I would use the `<details>` tag here, It makes the content collapsed by default, so people will just not recognise it. Is this something what is not important?

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 10]
```diff
@@ -0,0 +1,333 @@
...
+    }
+}
+// It shows the hierarchical structure along with the specific data types and corresponding values for each attribute
+```
+</details>
+
+### Flow-3.3: JSON to String (Go-Code)
+The SDK reads the JSON file containing the information about the Kubernetes resource and then translates this information into a string of Go code. This process involves parsing the JSON structure and generating corresponding Go code strings based on the structure, data types, and values extracted from the JSON representation. Ultimately, this results in a string that represents the Kubernetes resource in a format compatible with Go code.
+
+<details>
```
Comment (CsatariGergely) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:169]: I think this should be simply just normal description.

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+<summary>TraverseJSON Cases (Json-to-String)</summary>
+
+
+```
```
Comment (CsatariGergely) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:173]: Why is this intro text is added as a code example?

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 12]
```diff
@@ -0,0 +1,333 @@
...
+Based on the data type, Values are formatted accordingly,
+| Data-Type | Value    | Formatted-Value    |
+| :---:   | :---: | :---: |
+| int32 | 5   | 5   |
+| string | 5   | \"5\" |
+| *int32 | 5   | int32Ptr(5) |
+
+The Config-Jsons are required for more package-specific-types (such as : v1.Service, v1.Deployment)
+
+#### i) Struct_Module_mapping.json
```
Comment (CsatariGergely) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:275]:
```suggestion
#### Struct_Module_mapping.json
```

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 13]
```diff
@@ -0,0 +1,333 @@
...
+Mostly, It is seen that inspecting the type of struct(using reflect) would tell us that the struct belong to package “v1”, but there are multiple v1 packages (appsv1, metav1, rbacv1, etc), So, the actual package remains unknown. 
+
+Solution: To solve the above problems, we build a “structModuleMapping” which is a map that takes “struct-name” as the key and gives “package/module name” as a value.
+```
+v1.Deployment -->  appsv1.Deployment
+v1.Service --> corev1.Service
+```
+
+#### ii) Enum_Module_mapping.json
```
Comment (CsatariGergely) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:284]:
```suggestion
#### Enum_Module_mapping.json
```

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 14]
```diff
@@ -0,0 +1,333 @@
...
+
+Solution: We solve the above problems by building an “enumModuleMapping” which is a set that stores all data types that are enums. i.e. If a data type belongs to the set, then It is an Enum.
+
+There is an automation-script that takes the types.go files of packages and build the config-json. For details, Please refer [here](https://github.com/nephio-project/nephio-sdk/tree/main/helm-to-operator-codegen-sdk/config)
+
+
+### Flow-4: KRM to Unstruct-Obj to String(Go-code)
+All Kubernetes resource kinds that are not supported by the runtime-object method are handled using the unstructured method. In this approach, the Kubernetes Resource MOdel (KRM) is converted to an unstructured object using the package "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured". 
+Then, We traverse the unstructured-Obj in a DFS fashion and build the gocode-string.
+<details>
```
Comment (CsatariGergely) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:295]: I would leave the `<details>`

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 15]
```diff
@@ -0,0 +1,333 @@
...
+<summary>DFS Algorithm Cases (Unstruct-Version)</summary>
+
+
+```
```
Comment (CsatariGergely) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:299]: All of this is normal description and not example code.

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md [hunk 16]
```diff
@@ -0,0 +1,333 @@
+---
+title: Helm to Operator Codegen Sdk Developer Guide
```
Comment (CsatariGergely) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:2]: Is this a document for the developers of Helm to Operator Codegen Sdk?
If yes, then it should be moved to the Nephio contributor guides section.

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md [hunk 17]
```diff
@@ -0,0 +1,318 @@
...
+
+
+#### Example Run 
+```
+go run main.go /home/ubuntu/free5gccharts/towards5gs-helm/charts/free5gc/ free5gcns info
+```
+<details>
+<summary>The output is similar to:</summary>
+
+```console
```
Comment (CsatariGergely) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md:49]:
```suggestion
```bash
```

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md [hunk 18]
```diff
@@ -0,0 +1,318 @@
...
+After the basic structure of the operator is created, users can proceed to add their business logic
+The `CreateAll()` and `DeleteAll()` functions generated by the SDK can be leveraged for Day-0 resource deployments, allowing users to easily manage the creation and deletion of resources defined in the Helm chart. By integrating their business logic with these functions, users can ensure that their operator effectively handles resource lifecycle management and orchestration within a Kubernetes environment.
+
+In the end, all the resources created could be checked by:
+`kubectl get pods -n free5gcns`
+
+<details>
+<summary>The output Should be: </summary>
+
+```
```
Comment (CsatariGergely) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md:300]:
```suggestion
```bash
```

**Review by CsatariGergely — CHANGES_REQUESTED** [review]

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md (hunk 8 above)
Comment (jain-ashish-sam) [reply content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md:19]: Yes, I recall, there were some additional configurations while working with Calico CNI , but thanks for the info about Flannel, I will surely check it out!!

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md (hunk 16 above)
Comment (jain-ashish-sam) [reply content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md:2]: Yes, it is intended for developers, I will surely move it to  "Nephio contributor guides section."

**Comment by CsatariGergely** [comment]
> Hi @CsatariGergely , can you help me resolve the "broken-links" issue/ "linkspector job failing". I have checked the links in my own local hugo server, as well as the netlify-preview (https://deploy-preview-117--nephio.netlify.app/). The links seems to work there just fine. I have tried relative-links in one old-commit, but still the job-failed. Therefore, I am not able to get "why the job is failing." The logs of the job don't seem to be much helpful, (https://github.com/nephio-project/docs/actions/runs/8047908663/job/21978126772?pr=117)
> 
> ```
> Run linkspector check -c .linkspector.yml
>   linkspector check -c .linkspector.yml
>   shell: /usr/bin/bash -e {0}
> 💥 Error: Protocol error (Target.closeTarget): Session with given id not found.
> Error: Process completed with exit code 1.
> ```

Hi, I think it is "just" a flaky run. Someone with elevated rights could re-run the failed jobs.
I could run it in my ~~machine~~ fork and it gave more meaningful results: 
image

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md [hunk 19]
```diff
@@ -0,0 +1,318 @@
...
+release-name-free5gc-udm-udm-64d6c55855-6sl9n            1/1     Running             0          87m
+release-name-free5gc-udr-udr-7566b84c8b-5n7dv            1/1     Running             0          87m
+release-name-free5gc-upf-upf-76bd87cc76-bfqsw            1/1     Running             0          87m
+release-name-free5gc-webui-webui-7dccf6c877-lj5p5        1/1     Running             0          87m
+
+```
+</details>
+
+----
+For advanced requirements, Please refer to the developer guide [here](/docs/guides/contributor-guides/helm-to-operator-codegen-sdk-developer-guide.md)
```
Comment (CsatariGergely) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md:318]:
```suggestion
For advanced requirements, Please refer to the developer guide [here](/content/en/docs/guides/contributor-guides/helm-to-operator-codegen-sdk-developer-guide.md)
```

**Review by CsatariGergely — CHANGES_REQUESTED** [review]

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md (hunk 9 above)
Comment (CsatariGergely) [reply content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-developer-guide.md]: Yes, exactly :)

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md (hunk 19 above)
Comment (jain-ashish-sam) [reply content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md]: Thanks!!, The job passed finally!! :)

**Comment by jain-ashish-sam** [comment]
So, any more comments anyone?

**Comment by jain-ashish-sam** [comment]
@efiacor @lapentad @CsatariGergely @electrocucaracha @liamfallon @tliron ,If there not further comments, Can we merge this PR?

File: content/en/docs/guides/contributor-guides/helm-to-operator-codegen-sdk-developer-guide.md [hunk 20]
```diff
@@ -0,0 +1,328 @@
...
+2) Otherwise, if the resource is not Runtime-Supported, it is converted into an Unstructured Object and then into Go code.
+
+After the conversion process, all the generated Go code is gathered and compiled into a single Go file. This process results in a file that contains functions which can be readily utilized by Kubernetes operators.
+
+![alt Flow Diagram](/static/images/user-guides/helm-to-operator-codegen-sdk-flow-diagram.jpg)
+
+-----
+### Flow-1: Helm to YAML
+Helm to YAML conversion is achieved by running the following command
+`helm template <chart>  --namespace <namespace>  --output-dir “temp/templated/”` Internally. As of now, it retrieves the values from default "values.yaml"
```
Comment (electrocucaracha) [inline content/en/docs/guides/contributor-guides/helm-to-operator-codegen-sdk-developer-guide.md:20]:
```suggestion
`helm template <chart>  --namespace <namespace>  --output-dir “temp/templated/”` internally. As of now, it retrieves the values from default "values.yaml"
```

File: content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md [hunk 21]
```diff
@@ -0,0 +1,318 @@
...
+description: >
+  Deploying helm charts in Nephio using Helm To Operator Codegen Sdk
+weight: 1
+---
+
+The [Helm to Operator Codegen SDK](https://github.com/nephio-project/nephio-sdk/tree/main/helm-to-operator-codegen-sdk) offers a streamlined solution for translating existing Helm charts into Kubernetes operators with minimal effort and cost.
+
+By utilizing the Helm to Operator Codegen SDK, users can efficiently convert existing Helm charts into Kubernetes operators. This transition enables users to leverage the advanced capabilities provided by Kubernetes operators, such as enhanced automation, lifecycle management, and custom logic handling. Overall, the SDK facilitates a smooth migration process, empowering users to embrace the operator model for managing their Kubernetes resources effectively.
+
+## Excercise: Deploying Free5gc using operator
```
Comment (electrocucaracha) [inline content/en/docs/guides/user-guides/helm/helm-to-operator-codegen-sdk-user-guide.md:12]:
```suggestion
## Exercise: Deploying Free5gc using operator
```

**Comment by jain-ashish-sam** [comment]
@efiacor @lapentad @CsatariGergely @electrocucaracha @liamfallon @tliron , any further comments?

**Pull request merged**